README.md
# pybip39

`pybip39` is a fast Python library for
[BIP39](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki)
Bitcoin HD wallet mnemonic phrases. It supports multiple languages
and allows for seed phrases of 12 to 24 words. `pybip39` calls the Rust library
[tiny-bip39](https://github.com/maciejhirsz/tiny-bip39)
under the hood, thus benefitting from Rust's speed and safety.

[Online Docs](https://kevinheavey.github.io/pybip39/api_reference.html).

## Installation

`pip install pybip39`

Note: requires Python >= 3.7.

## Usage:

```python
from pybip39 import Mnemonic, Seed

mnemonic = Mnemonic()
# Get the phrase
phrase = mnemonic.phrase
print(f"phrase: {phrase}")
# Get the HD wallet seed
seed = Seed(mnemonic, "")
# get the HD wallet seed as raw bytes
seed_bytes = bytes(seed)
print(seed_bytes)

```

## Documentation credit

Most of this documentation is copied from
the [tiny-bip39 docs](https://docs.rs/tiny-bip39/latest/bip39/index.html).

## Development

### Setup

1. Install [poetry](https://python-poetry.org/)
2. Install dev dependencies:

```
poetry install
```

3. Activate the poetry shell:

```sh
poetry shell
```

### Testing

1. Run `maturin develop` to compile the Rust code.
2. Run `make fmt`, `make lint`, and `make test`.


CHANGELOG.md
## [0.1.0] - 2022-02-17

First release 🚀


tests.py
from pybip39 import Mnemonic


test_mnemonic.py
from pybip39 import Mnemonic


def test_back_to_back() -> None:
    m1 = Mnemonic(12, "en")
    m2 = Mnemonic.from_phrase(m1.phrase, "en")
    m3 = Mnemonic.from_entropy(m1.entropy, "en")
    assert m1.entropy == m2.entropy
    assert m1.entropy == m3.entropy
    assert m1.phrase == m2.phrase
    assert m1.phrase == m3.phrase


tests/test_seed.py
from pybip39 import Mnemonic, Seed, Language

ENTROPY = bytes(
    [
        0x33,
        0xE4,
        0x6B,
        0xB1,
        0x3A,
        0x74,
        0x6E,
        0xA4,
        0x1C,
        0xDD,
        0xE4,
        0x5C,
        0x90,
        0x84,
        0x6A,
        0x79,
    ]
)


def seed_hex_format() -> None:
    mnemonic = Mnemonic.from_entropy(ENTROPY)
    seed = Seed(mnemonic, "password")

    assert (
        seed.hex()
        == "0bde96f14c35a66235478e0c16c152fcaf6301e4d9a81d3febc50879fe7e5438e6a8dd3e39bdf3ab7b12d6b44218710e17d7a2844ee9633fab0e03d9a6c8569b"
    )


def check_unicode_normalization(
    lang: Language, phrase: str, password: str, expected_seed_hex: str
) -> None:
    mnemonic = Mnemonic.from_phrase(phrase, lang)
    seed = Seed(mnemonic, password)
    assert seed.hex() == expected_seed_hex


# Test vector is derived from https://github.com/infincia/bip39-rs/issues/26#issuecomment-586476647
def issue_26() -> None:
    check_unicode_normalization(
        Language.Spanish,
        "camello pomelo toque oponer urgente lástima merengue cutis tirón pudor pomo barco",
        "el español se habla en muchos países",
        "67a2cf87b9d110dd5210275fd4d7a107a0a0dd9446e02f3822f177365786ae440b8873693c88f732834af90785753d989a367f7094230901b204c567718ce6be",
    )


def test_password_is_unicode_normalized() -> None:
    check_unicode_normalization(
        Language.English,
        "abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about",
        "nullius　à　nym.zone ¹teſts² English",
        "61f3aa13adcf5f4b8661fc062501d67eca3a53fc0ed129076ad7a22983b6b5ed0e84e47b24cff23b7fca57e127f62f28c1584ed487872d4bfbc773257bdbc434",
    )


def test_japanese_normalization_1() -> None:
    check_unicode_normalization(
        Language.Japanese,
        "あいこくしん　あいこくしん　あいこくしん　あいこくしん　あいこくしん　あいこくしん　あいこくしん　あいこくしん　あいこくしん　あいこくしん　あいこくしん　あおぞら",
        "㍍ガバヴァぱばぐゞちぢ十人十色",
        "a262d6fb6122ecf45be09c50492b31f92e9beb7d9a845987a02cefda57a15f9c467a17872029a9e92299b5cbdf306e3a0ee620245cbd508959b6cb7ca637bd55",
    )


def test_japanese_normalization_2() -> None:
    check_unicode_normalization(
        Language.Japanese,
        "うちゅう　ふそく　ひしょ　がちょう　うけもつ　めいそう　みかん　そざい　いばる　うけとる　さんま　さこつ　おうさま　ぱんつ　しひょう　めした　たはつ　いちぶ　つうじょう　てさぎょう　きつね　みすえる　いりぐち　かめれおん",
        "㍍ガバヴァぱばぐゞちぢ十人十色",
        "346b7321d8c04f6f37b49fdf062a2fddc8e1bf8f1d33171b65074531ec546d1d3469974beccb1a09263440fc92e1042580a557fdce314e27ee4eabb25fa5e5fe",
    )


def test_french_normalization() -> None:
    check_unicode_normalization(
        Language.French,
        "paternel xénon curatif séparer docile capable exigence boulon styliste plexus surface embryon crayon gorge exister",
        "nullius　à　nym.zone ¹teſts² Français",
        "cff9ffd2b23549e73601db4129a334c81b28a40f0ee819b5d6a54c409999f0dfb6b89df17cae6408c96786165c205403d283baadc03ffdd391a490923b7d9493",
    )


def test_bytes() -> None:
    mnemonic = Mnemonic.from_entropy(ENTROPY)
    seed = Seed(mnemonic, "password")
    assert bytes(seed).hex() == seed.hex()


bench.py
import mnemonic
import pybip39


def test_pybip39_mnemonic(benchmark):
    benchmark(pybip39.Mnemonic)


def test_mnemonic_mnemonic(benchmark):
    benchmark(mnemonic.Mnemonic, "english")


def test_pybip39_words(benchmark):
    mnemo = pybip39.Mnemonic()
    benchmark(lambda: mnemo.phrase)


def test_mnemonic_words(benchmark):
    mnemo = mnemonic.Mnemonic("english")
    benchmark(mnemo.generate)


def _pybip39_mnemo_to_seed_bytes(mnemo: pybip39.Mnemonic) -> bytes:
    seed = pybip39.Seed(mnemo, "")
    return bytes(seed)


def _mnemonic_mnemo_to_seed_bytes(mnemo: mnemonic.Mnemonic) -> bytes:
    words = mnemo.generate()
    return mnemo.to_seed(words)


def test_pybip39_seed(benchmark):
    mnemo = pybip39.Mnemonic()
    benchmark(_pybip39_mnemo_to_seed_bytes, mnemo)


def test_mnemonic_seed(benchmark):
    mnemo = mnemonic.Mnemonic("english")
    benchmark(_mnemonic_mnemo_to_seed_bytes, mnemo)


src/lib.rs
use bip39::{
    Language as LanguageOriginal, Mnemonic as MnemonicOriginal,
    MnemonicType as MnemonicTypeOriginal, Seed as SeedOriginal,
};
use pyo3::prelude::*;

fn resolve_language(lang: &Language) -> LanguageOriginal {
    match lang {
        Language::English => LanguageOriginal::English,
        Language::ChineseSimplified => LanguageOriginal::ChineseSimplified,
        Language::ChineseTraditional => LanguageOriginal::ChineseTraditional,
        Language::French => LanguageOriginal::French,
        Language::Italian => LanguageOriginal::Italian,
        Language::Japanese => LanguageOriginal::Japanese,
        Language::Korean => LanguageOriginal::Korean,
        Language::Spanish => LanguageOriginal::Spanish,
    }
}

/// The primary type in this library - most tasks require creating or using one.
///
/// To create a *new* ``Mnemonic`` from a randomly generated key, call ``Mnemonic()``.
///
/// To get a ``Mnemonic`` instance for an existing mnemonic phrase, including
/// those generated by other software or hardware wallets, use :meth:`from_phrase`.
///
/// You can get the HD wallet seed from a ``Mnemonic`` by instantiating :class:`Seed`.
/// From there you can either get the raw byte value with ``bytes(seed)``, or the hex
/// representation with :meth:`Seed.hex`.
///
/// You can also get the original entropy value back from a ``Mnemonic`` with :attr:`entropy`,
/// but beware that the entropy value is **not the same thing** as an HD wallet seed, and should
/// *never* be used that way.
///
/// Args:
///     mtype (MnemonicType, optional): The number of words in the seed phrase. Defaults to :attr:`MnemonicType.Words12`.
///     language (Language, optional): The language of the seed phrase. Defaults to :attr:`Language.English`.
///
/// Example:
///      >>> from pybip39 import Mnemonic, MnemonicType, Language
///      >>> mnemonic = Mnemonic(MnemonicType.Words12, Language.English)
///      >>> phrase = mnemonic.phrase
///      >>> len(phrase.split(" "))
///      12
#[pyclass]
#[pyo3(text_signature = "(mtype, language)")]
struct Mnemonic(pub MnemonicOriginal);

#[pymethods]
impl Mnemonic {
    #[new]
    #[args(mtype = "MnemonicType::Words12", language = "Language::English")]
    pub fn new(mtype: MnemonicType, language: Language) -> Self {
        let mtype_original = match mtype {
            MnemonicType::Words12 => MnemonicTypeOriginal::Words12,
            MnemonicType::Words15 => MnemonicTypeOriginal::Words15,
            MnemonicType::Words18 => MnemonicTypeOriginal::Words18,
            MnemonicType::Words21 => MnemonicTypeOriginal::Words21,
            MnemonicType::Words24 => MnemonicTypeOriginal::Words24,
        };
        let resolved_language = resolve_language(&language);
        Self(MnemonicOriginal::new(mtype_original, resolved_language))
    }
    /// Create a :class:`Mnemonic` from pre-generated entropy.
    ///
    /// Args:
    ///     entropy (bytes): The entropy value of the mnemonic phrase.
    ///     language (Language, optional): The language of the seed phrase. Defaults to :attr:`Language.English`.
    ///
    /// Returns:
    ///     Mnemonic: The mnemonic object.
    ///
    /// Example:
    ///      >>> from pybip39 import Mnemonic, MnemonicType, Language
    ///      >>> entropy = bytes([0x33, 0xE4, 0x6B, 0xB1, 0x3A, 0x74, 0x6E, 0xA4, 0x1C, 0xDD, 0xE4, 0x5C, 0x90, 0x84, 0x6A, 0x79])
    ///      >>> mnemonic = Mnemonic.from_entropy(entropy, Language.English)
    ///      >>> mnemonic.phrase
    ///      'crop cash unable insane eight faith inflict route frame loud box vibrant'
    ///      >>> mnemonic.hex()
    ///      '33e46bb13a746ea41cdde45c90846a79'
    #[staticmethod]
    #[args(language = "Language::English")]
    #[pyo3(text_signature = "(entropy, language)")]
    pub fn from_entropy(entropy: &[u8], language: Language) -> PyResult<Self> {
        let lang = resolve_language(&language);
        let underlying = MnemonicOriginal::from_entropy(entropy, lang)?;
        Ok(Self(underlying))
    }
    /// Create a :class:`Mnemonic` from an existing mnemonic phrase.
    ///
    /// The phrase supplied will be checked for word length and validated according to the checksum
    /// specified in BIP0039
    ///
    /// Args:
    ///     phrase (str): The seed phrase.
    ///     language (Language, optional): The language of the seed phrase. Defaults to :attr:`Language.English`.
    ///
    /// Returns:
    ///     Mnemonic: The mnemonic object.
    ///
    /// Example:
    ///     >>> from pybip39 import Mnemonic, Language
    ///     >>> phrase = "park remain person kitchen mule spell knee armed position rail grid ankle"
    ///     >>> mnemonic = Mnemonic.from_phrase(phrase, Language.English)
    ///     >>> mnemonic.phrase
    ///     'park remain person kitchen mule spell knee armed position rail grid ankle'
    #[staticmethod]
    #[args(language = "Language::English")]
    #[pyo3(text_signature = "(phrase, language)")]
    pub fn from_phrase(phrase: &str, language: Language) -> PyResult<Self> {
        let lang = resolve_language(&language);
        let underlying = MnemonicOriginal::from_phrase(phrase, lang)?;
        Ok(Self(underlying))
    }

    /// Validate a mnemonic phrase.
    ///
    /// The phrase supplied will be checked for word length and validated according to the checksum
    /// specified in BIP0039.
    ///
    /// Returns ``None`` if phrase is valid.
    ///
    /// Args:
    ///     phrase (str): The phrase to validate.
    ///     language (Language, optional): The language of the phrase. Defaults to :attr:`Language.English`.
    ///
    /// Raises:
    ///      RuntimeError: If validation fails.
    ///
    /// Example:
    ///      >>> from pybip39 import Mnemonic, Language
    ///      >>> test_mnemonic = "park remain person kitchen mule spell knee armed position rail grid ankle"
    ///      >>> Mnemonic.validate(test_mnemonic, Language.English)
    ///      >>> Mnemonic.validate(test_mnemonic, Language.Spanish)
    ///      Traceback (most recent call last):
    ///        File "<stdin>", line 1, in <module>
    ///      RuntimeError: invalid word in phrase
    #[staticmethod]
    #[args(language = "Language::English")]
    #[pyo3(text_signature = "(phrase, language)")]
    pub fn validate(phrase: &str, language: Language) -> PyResult<()> {
        let lang = resolve_language(&language);
        Ok(MnemonicOriginal::validate(phrase, lang)?)
    }
    /// bytes: The original entropy value of the mnemonic phrase.
    ///
    /// Example:
    ///     >>> from pybip39 import Mnemonic
    ///     >>> phrase = "park remain person kitchen mule spell knee armed position rail grid ankle"
    ///     >>> mnemonic = Mnemonic.from_phrase(phrase)
    ///     >>> mnemonic.entropy
    ///     b'\xa0V\xb2\x8d=\x89\x15\xa2^\xe0^\xa8v\x1d\x99\x84'
    ///
    /// Note:
    ///     You shouldn't use the generated entropy as secrets, for that generate a new `Seed` from the `Mnemonic`.
    #[getter]
    pub fn entropy(&self) -> &[u8] {
        self.0.entropy()
    }

    /// str: The mnemonic phrase.
    #[getter]
    pub fn phrase(&self) -> &str {
        self.0.phrase()
    }

    /// Get the hex-formatted mnemonic.
    ///
    /// Returns:
    ///      str: The hex-formatted mnemonic.
    pub fn hex(&self) -> String {
        format!("{:x}", self.0)
    }

    pub fn __str__(&self) -> &str {
        self.phrase()
    }

    pub fn __bytes__(&self) -> &[u8] {
        self.entropy()
    }
}

/// The supported languages for seed phrases.
#[pyclass]
#[derive(Clone)]
enum Language {
    English,
    ChineseSimplified,
    ChineseTraditional,
    French,
    Italian,
    Japanese,
    Korean,
    Spanish,
}

/// The supported word counts for seed phrases.
#[pyclass]
#[derive(Clone)]
pub enum MnemonicType {
    Words12,
    Words15,
    Words18,
    Words21,
    Words24,
}

/// The secret value used to derive HD wallet addresses from a :class:`Mnemonic` phrase.
///
/// Because it is not possible to create a :class:`Mnemonic` instance that is invalid, it is
/// therefore impossible to have a ``Seed`` instance that is invalid. This guarantees that only
/// a valid, intact mnemonic phrase can be used to derive HD wallet addresses.
///
/// To get the raw byte value use ``bytes(seed)``. These can be used to derive
/// HD wallet addresses using another crate (deriving HD wallet addresses is outside the scope of this
/// crate and the BIP39 standard).
///
/// Args:
///      mnemonic (Mnemonic): The mnemonic to generate the seed from.
///      password (str): The seed password.
#[pyclass]
#[pyo3(text_signature = "(mnemonic, password)")]
struct Seed(pub SeedOriginal);

#[pymethods]
impl Seed {
    #[new]
    pub fn new(mnemonic: &Mnemonic, password: &str) -> Self {
        Self(SeedOriginal::new(&mnemonic.0, password))
    }

    pub fn __bytes__(&self) -> &[u8] {
        self.0.as_bytes()
    }

    /// Get the hex-formatted seed.
    ///
    /// Returns:
    ///      str: The hex-formatted seed.
    pub fn hex(&self) -> String {
        format!("{:x}", self.0)
    }
}

#[pymodule]
fn pybip39(_py: Python, m: &PyModule) -> PyResult<()> {
    m.add_class::<Mnemonic>()?;
    m.add_class::<Language>()?;
    m.add_class::<Seed>()?;
    m.add_class::<MnemonicType>()?;
    Ok(())
}
